installer/scripts/service_control

```diff
@@ -69,49 +176,107 @@ verify_privileges()
     fi
 }
 
-is_omsagent_running()
+this_omsagent_running()
 {
-    verify_privileges
+    if [ -f $PIDFILE ]; then
+        ps -p `cat $PIDFILE` | grep -q omsagent
+        WS_STATUS=$?
+    else
+        WS_STATUS=$FILE_NOT_FOUND
+    fi
+    return $WS_STATUS
+}
 
+is_omsagent_running()
+{
     # Returns 1 if 'omsagent' is running, 0 otherwise
-    [ -f $PIDFILE ] || return 0
-    ps -p `cat $PIDFILE` | grep -q omsagent
-    STATUS=$?
+    #### Keeping return value on this as it originally was.  I deduce
+    #### that because this value is returned with the is-running qualifier
+    #### to service_control, that there are scripts in other languages
+    #### that use the more conventional values for booleans.  Meanwhile,
+    #### I recommend deprecation in all but the external interface.  XC
+    if this_omsagent_running; then
+        return 1
+    else
+        return 0
+    fi
+} 
+
+remove_PIDfile_unless_omsagent_running()
+{
 
-    # Kill PID file if we're not running any longer
-    if [ $STATUS -ne 0 ]; then
-        rm -f $PIDFILE
+    WS_STATUS=0
+    if this_omsagent_running; then
         return 0
-    else
-        return 1
     fi
+    rm -f $PIDFILE # NOTE:  definition of PIDFILE is dependent on WS.
+    if [ $? -ne 0 ]; then
+        WS_STATUS=2
+        return $WS_STATUS
+    fi
+    return 1
 }
 
 exit_if_agent_not_onboarded()
 {
     if [ ! -f $OMSADMIN_CONF ]; then
-        echo "Warning: Agent is not onboarded. omsagent cannot be registered as a service."
-        exit 0 
+        echo "FATAL: Agent is not onboarded. omsagent cannot be registered as a service."
+        WS_STATUS=$FILE_NOT_FOUND
+        exit $WS_STATUS
     fi
 }
 
+wait_until_omsagent_is_running()
+{
+    # Required parameter: Number of seconds to wait for agent to become running.
+    if [ -z "$1" -o "$1" -le 0 ]; then
+        echo "Function \"wait_until_omsagent_stops\" called with invalid parameter"
+        WS_STATUS=1
+        exit $WS_STATUS
+    fi
+
+    COUNTER=$(( $1 * 2 )) # Since we sleep 0.5 seconds, compute number of seconds
+    while [ $COUNTER -gt 0 ]; do
+        if ! this_omsagent_running; then
+            COUNTER=$(( $COUNTER - 1 ))
+            sleep 0.5
+        else
+            WS_STATUS=0
+            return $WS_STATUS
+        fi
+    done
+
+    # One final try for accurate return status (just return status from the call)
+    remove_PIDfile_unless_omsagent_running
+    WS_STATUS=$? # Should only be non-zero of the removal of a PIDFILE failed for some reason.
```

Comment:
As I mentioned before, this comment indicates that `remove_PIDfile_unless_omsagent_running` only returns non-zero if the PIDFILE was attempted to be removed, but that failed. This contradicts [your code](https://github.com/Microsoft/OMS-Agent-for-Linux/blob/7dc009e88cea05af7603cda59c2c6a4aa1700a5e/installer/scripts/service_control#L213), which reads:
```
    rm -f $PIDFILE # NOTE:  definition of PIDFILE is dependent on WS.
    if [ $? -ne 0 ]; then
        WS_STATUS=2
        return $WS_STATUS
    fi
    return 1
```
If the `rm -f $PIDFILE` line returns 0, you return 1 from `remove_PIDfile_unless_omsagent_running`! If you want to use that method as a trinary state return (0 for omsagent_running, 1 for omsagent not running and PIDFILE removed, 2 for omsagent not running and PIDFILE not removed, than you need to change the rest of your code that uses its return value.